insightsengineering/teal
# Fixes lingering/ghost slices in the filter panel after DDL reload

# Pull Request

#### Fixes
- teal.slices - after removal of the slice make sure that observers of FilterState are destroyed. To test - take an app with DDL. Add filter and reload data several times. After this try to remove filter with remove button. What is interesting that single FIlterState card remove button needs to be pushed several times (it removes old FIlterStates which are not rendered anymore) while remove-all seems to not add observers one over another (it is enough to click once).


#### Needs

Branch `teal_slice@669_insertUI@main` on `{teal.slices}`

#### Changes description

- Keeps previous FilteredDataset to safely clean the filter panel

##### How to test?

1. Run app below
2. Reload the dataset using the "`data` tab"
3. Remove individual filter from filter panel
4. Check if filter has been removed with a single click

Repeat test without `module_specific = TRUE`

```r
# teal.slices -
# after removal of the slice make sure that observers of FilterState are destroyed.
#
# To test - take an app with DDL.
# - Add filter
# - reload data several times.
#
# After this try to remove filter with remove button.
# What is interesting that single FIlterState card remove button needs to be
# pushed several times (it removes old FIlterStates which are not rendered anymore)
# while remove-all seems to not add observers one over another (it is enough to
# click once). @averissimo

# Possible bug: reloading data will clear default values of the initial filter state

options(
  teal.log_level = "DEBUG",
  teal.show_js_log = TRUE,
  # teal.bs_theme = bslib::bs_theme(version = 5),
  shiny.bookmarkStore = "server"
)

pkgload::load_all("../teal.slice")
pkgload::load_all("../teal")

data <- teal::teal_data_module(
  ui = function(id) {
    ns <- shiny::NS(id)
    shiny::tagList(
      shiny::textInput(ns("username"), label = "Username"),
      shiny::passwordInput(ns("password"), label = "Password"),
      shiny::actionButton(ns("submit"), label = "Submit")
    )
  },
  server = function(id, ...) {
    shiny::moduleServer(id, function(input, output, session) {
      shiny::eventReactive(input$submit, {
        data <- teal.data::teal_data() |>
          within(
            {
              logger::log_debug("Loading data...")
              ADSL <- teal.data::rADSL
              ADTTE <- teal.data::rADTTE
              iris <- iris

              CO2 <- CO2
              factors <- names(Filter(isTRUE, vapply(CO2, is.factor, logical(1L))))
              CO2[factors] <- lapply(CO2[factors], as.character)
            },
            username = input$username,
            password = input$password
          )
        join_keys(data) <- default_cdisc_join_keys[c("ADSL", "ADTTE")]
        teal.data::datanames(data) <- c("ADSL", "ADTTE", "iris", "CO2")
        data
      })
    })
  }
)

teal::init(
  data = data,
  modules = teal::modules(
    teal::modules(
      label = "Group",
      teal::example_module()
    )
  ),
  filter = teal::teal_slices(
    teal.slice::teal_slice("ADSL", "AGE", selected = c(18L, 65L)),
    module_specific = TRUE,
    teal.slice::teal_slice("ADSL", "SEX")
  ),
  title = "yada"
) |>
  shiny::runApp()
```
